core/go/internal/sequencer/coordinator/transaction/state_machine.go

```diff
@@ -0,0 +1,556 @@
+	Event_AssembleTimeout                                                                // the assemble timeout period has passed since we sent the first assemble request
+)
+
+type StateMachine struct {
+	currentState    State
+	lastStateChange time.Time
+}
+
+// Actions can be specified for transition to a state either as the OnTransitionTo function that will run for all transitions to that state or as the On field in the Transition struct if the action applies
+// for a specific transition
+type Action func(ctx context.Context, txn *Transaction) error
+
+// TODO should we pass static config ( e.g. timeouts) to the guards instead of storing them in every transaction struct instance?
+type Guard func(ctx context.Context, txn *Transaction) bool
+
+type Transition struct {
+	To State // State to transition to if the guard condition is met
+	If Guard // Condition to evaluate the transaction against to determine if this transition should be taken
+	On Action
+}
+
+type ActionRule struct {
+	Action Action
+	If     Guard
+}
+
+type EventHandler struct {
+	Validator   func(ctx context.Context, txn *Transaction, event common.Event) (bool, error) // function to validate whether the event is valid for the current state of the transaction.  This is optional.  If not defined, the event is always considered valid.
+	Actions     []ActionRule                                                                  // list of actions to be taken when this event is received.  These actions are run before any transition specific actions
+	Transitions []Transition                                                                  // list of transitions that this event could trigger.  The list is ordered so the first matching transition is the one that will be taken.
+}
+
+type StateDefinition struct {
+	OnTransitionTo Action                     // function to be invoked when transitioning into this state.  This is invoked after any transition specific actions have been invoked
+	Events         map[EventType]EventHandler // rules to define what events apply to this state and what transitions they trigger.  Any events not in this list are ignored while in this state.
+}
+
+var stateDefinitionsMap map[State]StateDefinition
+
+func init() {
+	// Initialize state definitions in init function to avoid circular dependencies
+	stateDefinitionsMap = map[State]StateDefinition{
+		State_Initial: {
+			Events: map[EventType]EventHandler{
+				Event_Received: { //TODO rename this event type because it is the first one we see in this struct and it seems like we are saying this is a definition related to receiving an event (at one level that is correct but it is not what is meant by Event_Received)
+					Transitions: []Transition{
+						{
+							To: State_Pooled,
+							If: guard_And(guard_Not(guard_HasUnassembledDependencies), guard_Not(guard_HasUnknownDependencies)),
```

Comment:
Unclear to me what event changes `HasUnassembledDependencies`, and which states you are allowed to be in where it is valid for dependencies to be changed.